File: Products/sample-android/app/src/androidTest/java/mockifer/example/RecyclerViewCountAssertion.java
package mockifer.example;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

import android.support.test.espresso.NoMatchingViewException;
import android.support.test.espresso.ViewAssertion;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import org.hamcrest.Matcher;

class RecyclerViewCountAssertion implements ViewAssertion {
    private final Matcher<Integer> matcher;

    @Override public void check(View view, NoMatchingViewException noViewFoundException) {
        if (noViewFoundException != null) {
            throw noViewFoundException;
        }

        RecyclerView recyclerView = (RecyclerView) view;
        RecyclerView.Adapter adapter = recyclerView.getAdapter();
        assertThat(adapter.getItemCount(), matcher);
    }

    static RecyclerViewCountAssertion withItemCount(int expectedCount) {
        return withItemCount(is(expectedCount));
    }

    static RecyclerViewCountAssertion withItemCount(Matcher<Integer> matcher) {
        return new RecyclerViewCountAssertion(matcher);
    }

    private RecyclerViewCountAssertion(Matcher<Integer> matcher) {
        this.matcher = matcher;
    }
}

File: support/samples/sample-android/app/src/androidTest/java/mockifer/example/RecyclerViewCountAssertion.kt
package mockifer.example

import android.support.test.espresso.NoMatchingViewException
import android.support.test.espresso.ViewAssertion
import android.support.v7.widget.RecyclerView
import android.view.View
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.Matcher
import org.hamcrest.MatcherAssert.assertThat

internal class RecyclerViewCountAssertion private constructor(private val matcher: Matcher<Int>) : ViewAssertion {

    override fun check(view: View, noViewFoundException: NoMatchingViewException?) {
        if (noViewFoundException != null) {
            throw noViewFoundException
        }

        val recyclerView = view as RecyclerView
        val adapter = recyclerView.adapter
        assertThat(adapter.itemCount, matcher)
    }

    companion object {

        fun withItemCount(expectedCount: Int): RecyclerViewCountAssertion {
            return withItemCount(`is`(expectedCount))
        }

        fun withItemCount(matcher: Matcher<Int>): RecyclerViewCountAssertion {
            return RecyclerViewCountAssertion(matcher)
        }
    }
}

